lib/postgrex/protocol.ex

```diff
@@ -787,151 +786,333 @@ defmodule Postgrex.Protocol do
+            msg_sync()]
+
+    with :ok <- msg_send(%{s | buffer: nil}, msgs, buffer),
+         {:ok, s, buffer} <- recv_close(s, status, buffer),
+         query_delete(s, query),
+         {:ok, query, s, buffer}
+          <- recv_parse_describe(s, status, query, buffer),
+         {:ok, s} <- recv_ready(s, status, buffer) do
+      {:ok, query, s}
+    else
+      {:reload, oid, s, buffer} ->
+        reload_ready(s, status, query, oid, buffer)
+      {:disconnect, err, s} ->
+        {:disconnect, err, s}
+      {:error, %Postgrex.Error{} = err, s, buffer} ->
+        error_ready(s, status, err, buffer)
+    end
   end
 
-  defp close_parse_describe(s, status, query, buffer, next) do
+  defp close_parse_describe_flush(s, %{mode: :transaction} = status, query) do
+    %Query{name: name, statement: statement} = query
+    msgs = [msg_close(type: :statement, name: name),
+            msg_parse(name: name, statement: statement, type_oids: []),
+            msg_describe(type: :statement, name: name),
+            msg_flush()]
+
+    %{buffer: buffer} = s
+    with :ok <- msg_send(%{s | buffer: nil}, msgs, buffer),
+         {:ok, s, buffer} <- recv_close(s, status, buffer),
+         query_delete(s, query),
+         {:ok, %Query{ref: ref} = query, %{postgres: postgres} = s, buffer}
+          <- recv_parse_describe(s, status, query, buffer) do
+      # lock state with unique query reference as not synced
+      {:ok, query, %{s | postgres: {postgres, ref}, buffer: buffer}}
+    else
+      {:error, err, s, buffer} ->
+        error_flushed(s, status, err, buffer)
+      {:reload, oid, s, buffer} ->
+        reload_flushed(s, status, query, oid, buffer)
+      {:disconnect, _err, _s} = disconnect ->
+        disconnect
+    end
+  end
+  defp close_parse_describe_flush(%{postgres: :transaction, buffer: buffer} = s,
+       %{mode: :savepoint} = status, query) do
     %Query{name: name, statement: statement} = query
-    msgs =
-      [msg_close(type: :statement, name: name),
-       msg_parse(name: name, statement: statement, type_oids: []),
-       msg_describe(type: :statement, name: name)]
-    describe_recv = &describe_recv(&1, &2, &3, &4, next)
-    parse_recv = &parse_recv(&1, &2, &3, &4, describe_recv)
-    recv = &close_recv(&1, &2, &3, &4, parse_recv)
-    send_and_recv(s, status, query, buffer, msgs, recv)
+    msgs = [msg_query(statement: "SAVEPOINT postgrex_query"),
+            msg_close(type: :statement, name: name),
+            msg_parse(name: name, statement: statement, type_oids: []),
+            msg_describe(type: :statement, name: name),
+            msg_flush()]
+
+    with :ok <- msg_send(%{s | buffer: nil}, msgs, buffer),
+         {:ok, _, %{buffer: buffer} = s} <- recv_transaction(s, status, buffer),
+         {:ok, s, buffer} <- recv_close(s, status, buffer),
+         {:ok, %Query{ref: ref} = query, %{postgres: postgres} = s, buffer}
+          <- recv_parse_describe(s, status, query, buffer) do
```

Comment:
Weird line break here.

Reply:
Can you expand?

Reply:
Maybe the <- match should be split into multiple matches if it cant fit a single line.

Reply:
The code was written before the formatter existed. I'll run the formatter on it and try to get it formatting nicely.

Reply:
I updated the `with` clauses to use the formatter style, seems both indentation and `<-` placement were not quite right in many cases. I didn't move pattern matching into multiple lines because when it was removed the lines still needed multilines.